### PR TITLE
Add page on `magicgui` type registration

### PR DESCRIPTION
**Pull request by lucyleeow**
# References and relevant issues
Came about after #68

Details `magicgui` type registration, in particular the `reset_choices` connection to layers we do for all widgets that have that method.

### CODE REVIEW

**Comment by lucyleeow** [comment]
cc @DragaDoncila

**Review by DragaDoncila — DISMISSED** [review]
LGTM

**Review by DragaDoncila — COMMENTED** [review]
Umm sorry, when I approved I think I'd only loaded one commit and this PR seemed much smaller... Let me actually go through this sorry

File: docs/developers/architecture/index.md [hunk 1]
```diff
@@ -11,3 +11,6 @@ code base. For advanced napari usage documentation, see [](explanations).
   connected to Qt classes and Vispy classes.
 - [](app-model): Explains the napari application model, a declarative schema for
   keeping track of commands, menus and keybindings of the napari GUI.
+- [](magicgui_type_registration): Explains how `magicgui` widgets are
```
Comment (DragaDoncila) [inline docs/developers/architecture/index.md]:
```suggestion
- [](magicgui_type_registration): Explains how `magicgui` widgets declared by users or plugins are
```

File: docs/developers/architecture/magicgui_type_reg.md [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+```
+
+## Registration details
+
+`napari` types are either registered via the
+{func}`@register_type <magicgui.type_map.register_type>` decorator when they are defined or in
+[`napari/types.py`](https://github.com/napari/napari/blob/main/napari/types.py).
+For the full list of types registered, see [](magicgui-parameter-annotations).
+
+All 'layer' types provide a `choices` callable when registering.
```
Comment (DragaDoncila) [inline docs/developers/architecture/magicgui_type_reg.md]:
```suggestion
All 'layer' types provide a `choices` callable when they are registered with `magicgui`.
```

File: docs/developers/architecture/magicgui_type_reg.md [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+whenever layers change.
+**Note that {meth}`~napari.qt.Window.add_dock_widget` will
+connect any existing `reset_choices` widget attribute to layer events for all widgets,
+not just `magicgui` widgets.**
+
+The 'layer' types also specify a `return_callback` function that adds the layer
+to the closest parent `Viewer` of the native widget when a 'layer' type is a return
+annotation.
+
+{class}`~napari.viewer.Viewer` differs from the Layer types. `napari` simply specifies
```
Comment (DragaDoncila) [inline docs/developers/architecture/magicgui_type_reg.md]: Is the `Viewer` registered with magicgui at all? If not then I think we should explicitly state that here

File: docs/developers/architecture/magicgui_type_reg.md [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+that the {class}`~napari.viewer.Viewer` be bound to a widget (technically a
+hidden {class}`~magicgui.widgets.EmptyWidget`). The user will need to specify
+the widget type for this annotation.
+
+```{important}
+`magicgui` type registration allows `napari` to specify useful defaults when creating
```
Comment (DragaDoncila) [inline docs/developers/architecture/magicgui_type_reg.md]: I am confused about what this is trying to say sorry

**Review by DragaDoncila — COMMENTED** [review]
Ok left some comments @lucyleeow  🙏

File: docs/developers/architecture/magicgui_type_reg.md (hunk 3 above)
Comment (lucyleeow) [reply docs/developers/architecture/magicgui_type_reg.md]: It is registered and it is super weird ([ref](https://github.com/napari/napari/blob/cf68ee5ffbe6f5ff4861642c759eaff181ef6749/napari/viewer.py#L15) ):
```python
@mgui.register_type(bind=_magicgui.proxy_viewer_ancestor)
class Viewer(ViewerModel):
```

what 'bind' does ([ref](https://github.com/pyapp-kit/magicgui/blob/1e78e6f0bee37aeb9ce7d300a52f5f120141f834/src/magicgui/type_map/_type_map.py#L423C1-L426C68)):

```python
    elif "bind" in _options:
        # if we're binding a value to this parameter, it doesn't matter what type
        # of ValueWidget is used... it usually won't be shown
        _TYPE_DEFS[resolved_type] = (widgets.EmptyWidget, _options)
```

`EmptyWidget` is just a 'hidden' widget, the user has to specify widget they want for `Viewer` annotation. I don't know how a `Viewer` would be used as an input widget though..?!

File: docs/developers/architecture/magicgui_type_reg.md (hunk 4 above)
Comment (lucyleeow) [reply docs/developers/architecture/magicgui_type_reg.md]: Welp, I got confused. Removed this.

**Comment by lucyleeow** [comment]
Thank you @DragaDoncila! I've updated

File: docs/developers/architecture/magicgui_type_reg.md [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+## Registration details
+
+`napari` types are either registered via the
+{func}`@register_type <magicgui.type_map.register_type>` decorator when they are
+defined or in
+[`napari/types.py`](https://github.com/napari/napari/blob/main/napari/types.py).
+For the full list of types registered, see [](magicgui-parameter-annotations).
+
+All 'layer' types provide a `choices` callable when they are registered with
+`magicgui`. This means that annotating with these types creates an
```
Comment (DragaDoncila) [inline docs/developers/architecture/magicgui_type_reg.md]:
```suggestion
`magicgui`. This means that annotating with these types creates a
```

File: docs/developers/architecture/magicgui_type_reg.md [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+{class}`~magicgui.widgets.bases.CategoricalWidget`, which will have a dropdown
+selection will get updated via the
```
Comment (DragaDoncila) [inline docs/developers/architecture/magicgui_type_reg.md]:
```suggestion
selection whose options will be updated via the
```

File: docs/developers/architecture/magicgui_type_reg.md [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+`choices` callable. This callable is either `get_layers_data` or `get_layers`.
+These functions retrieve the closest parent `Viewer` of the native
+{class}`~magicgui.widgets.bases.CategoricalWidget` widget and returns a list of
+{class}`~napari.layers.Layer` or tuple of format ('layer name', `<LayerType>Data`).
```
Comment (DragaDoncila) [inline docs/developers/architecture/magicgui_type_reg.md]:
```suggestion
{class}`~napari.layers.Layer` or tuple of format `(layer name, <LayerType>Data)`.
```

File: docs/developers/architecture/magicgui_type_reg.md [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+
+```{note}
+This page provides implementation details on `napari`-specific type registration
+in `magicgui` and is aimed at `napari` developers.
+
+For information about using `magicgui` (for users and plugin developers) see
+[](creating-widgets).
+```
+
+## Registration details
```
Comment (DragaDoncila) [inline docs/developers/architecture/magicgui_type_reg.md:26]: Would be good to have some subheadings here maybe? One for `layers` and one for `viewer` at least?

**Review by DragaDoncila — COMMENTED** [review]
@lucyleeow sorry, couple more comments. I think this is looking good!

**Comment by lucyleeow** [comment]
Thanks @DragaDoncila, changes made!

File: docs/developers/architecture/magicgui_type_reg.md [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+
+### Layer types
+
+The 'layer' types registered are {class}`~napari.layers.Layer` and its subclasses, and
+`<LayerType>Data` types.
+All these types provide a `choices` callable when they are registered with
+`magicgui`. This means that annotating with these types creates a
+{class}`~magicgui.widgets.bases.CategoricalWidget`, which will have a dropdown
+selection whose options will be updated via the
+`choices` callable. This callable is either `get_layers_data` or `get_layers`.
```
Comment (DragaDoncila) [inline docs/developers/architecture/magicgui_type_reg.md]:
```suggestion
`choices` callable. 

This callable is either `get_layers_data` or `get_layers`.
```

File: docs/developers/architecture/magicgui_type_reg.md [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+These functions retrieve the closest parent `Viewer` of the native
+{class}`~magicgui.widgets.bases.CategoricalWidget` widget and returns a list of
+{class}`~napari.layers.Layer` or tuple of format `('layer name', <LayerType>Data)`.
+This callable is set to the `choices` attribute of the
+{class}`~magicgui.widgets.bases.CategoricalWidget` in its
+{meth}`~magicgui.widgets.bases.CategoricalWidget.reset_choices` method and thus
+gets called via {class}`~magicgui.widgets.bases.CategoricalWidget`'s `choices` setter.
+`napari` {meth}`~napari.qt.Window.add_dock_widget` checks if the dock widget has
+a `reset_choices` attribute and if so, connects it to layer events.
+Note that `magicgui` {class}`~magicgui.widgets.bases.ContainerWidget`'s will call
```
Comment (DragaDoncila) [inline docs/developers/architecture/magicgui_type_reg.md:55]:
```suggestion

Note that `magicgui` {class}`~magicgui.widgets.bases.ContainerWidget`'s will call
```

**Review by DragaDoncila — APPROVED** [review]
@lucyleeow approved with just two more comments to break up a big paragraph (looks a bit intense on the built docs otherwise)

**Pull request merged**